src/app.ts

```diff
@@ -117,44 +106,154 @@ async function runWizard() {
+      await removeExcludedFilesAndAddTokenCount(fileTree);
+
+    if (!completedTreeWork || completedTreeWork === "timeoutFailed" || !fileTree.children) {
+      console.log(
+        completedTreeWork && completedTreeWork !== "timeoutFailed"
+          ? "\nNo files found in directory. Try again."
+          : "\nThe file tree is too large to process in a reasonable time."
+      );
+      return;
+    }
+
+    fileTree = completedTreeWork.tree;
+    let first_time = true;
+    let selectedFiles: string[] = [];
+    let file_choices: { name: string; value: string; checked: boolean }[] | "timeoutFailed" = await flattenFileTreeForCheckbox(fileTree);
+
+    if (!file_choices || file_choices === "timeoutFailed") {
+      console.log("\nThe file tree is too large to process in a reasonable time. Exiting.");
+      return; // Return if the timeout is reached
+    }
+
+    let promptTokens = getAdditionalPromptTokens(file_choices);
+
+    // Loop until the files selected are within the token limit
+    while (first_time || completedTreeWork.tokenTotal + promptTokens > CLAUDE_PRIMARYSOURCE_BUDGET) {
+      if (!first_time) {
+        console.log("\nYou've selected too many tokens. Please deselect files to exclude.");
+      }
+      first_time = false;
+
+      console.log("before checkbox await");
+
+      selectedFiles = await checkbox({
+        pageSize: 8,
+        loop: false,
+        message: `The token limit is ${CLAUDE_PRIMARYSOURCE_BUDGET.toLocaleString()}. 
+Your current file token count is ${completedTreeWork.tokenTotal.toLocaleString()}, with ${promptTokens.toLocaleString()} for the prompt, for a total of ${(
+            completedTreeWork.tokenTotal + promptTokens
+          ).toLocaleString()}.
+Please deselect files to exclude.
+Note: If you deselect a folder, all files within it will be excluded.
+Note: Some files do not appear as we don't believe we can read them.`,
+        choices: file_choices,
+        theme: { style: { renderSelectedChoices: () => { } } }
+      });
+
+      console.log("after checkbox await");
+
+      completedTreeWork = await removeDeselectedItems(fileTree, selectedFiles);
+      if (!completedTreeWork || completedTreeWork === "timeoutFailed" || !fileTree.children) {
+        console.log(
+          completedTreeWork && completedTreeWork !== "timeoutFailed"
+            ? "No files found in directory. Try again."
+            : "The file tree is too large to process in a reasonable time."
+        );
+        return;
+      }
+      fileTree = completedTreeWork.tree;
+      file_choices = await flattenFileTreeForCheckbox(fileTree);
+      if (!file_choices || file_choices === "timeoutFailed") {
+        console.log("The file tree is too large to process in a reasonable time. Exiting.");
+        return; // Return if the timeout is reached
+      }
+      promptTokens = getAdditionalPromptTokens(file_choices);
+    }
```

Comment:
Should we move this whole section out to a new function to keep the wizard processing section simpler?

Reply:
The wizard processing section was already long enough that I didn't think this added too much complexity - and there's also plenty references to `wizard` state in this section. I'd thought the generat structure was that all references to `wizard` state were to be in the `app.ts`. 

If we want to tidy up `app.ts` maybe that should be a separate PR? I tried a bit here by adding code comments to separate out the different sections.